scrapy_poet/injection.py

```diff
@@ -0,0 +1,316 @@
+from scrapy.http import Response
+from scrapy.settings import Settings
+from scrapy.statscollectors import StatsCollector
+from scrapy.utils.conf import build_component_list
+from scrapy.utils.defer import maybeDeferred_coro
+from scrapy.utils.misc import load_object
+from scrapy_poet.injection_errors import (UndeclaredProvidedTypeError,
+                                          NonCallableProviderError,
+                                          InjectionError)
+from scrapy_poet.page_input_providers import PageObjectInputProvider
+from scrapy_poet.api import _CALLBACK_FOR_MARKER, DummyResponse
+from web_poet.pages import is_injectable
+
+
+logger = logging.getLogger(__name__)
+
+
+class Injector:
+    """
+    Keep all the logic required to do dependency injection in Scrapy callbacks.
+    Initializes the providers from the spider settings at initialization.
+    """
+    def __init__(self, crawler: Crawler, default_providers: Mapping = None):
+        self.crawler = crawler
+        self.spider = crawler.spider
+        self.load_providers(default_providers)
+
+    def load_providers(self, default_providers: Mapping = None):
+        providers_dict = {**(default_providers or {}),
+                          **self.spider.settings.getdict("SCRAPY_POET_PROVIDERS")}
+        provider_classes = build_component_list(providers_dict)
+        logger.info(f"Loading providers {provider_classes}")
+        self.providers = [
+            load_object(cls)(self.crawler)
+            for cls in provider_classes
+        ]
+        check_all_providers_are_callable(self.providers)
+        # Caching whether each provider requires the scrapy response
+        self.is_provider_requiring_scrapy_response = {
+            provider: is_provider_requiring_scrapy_response(provider)
+            for provider in self.providers
+        }
+        # Caching the function for faster execution
+        self.is_class_provided_by_any_provider = \
+            is_class_provided_by_any_provider_fn(self.providers)
+
+    def available_dependencies_for_providers(self,
+                                             request: Request,
+                                             response: Response):
+        deps = {
+            Crawler: self.crawler,
+            Spider: self.spider,
+            Settings: self.spider.settings,
+            StatsCollector: self.crawler.stats,
+            Request: request,
+            Response: response,
+        }
+        assert deps.keys() == SCRAPY_PROVIDED_CLASSES
+        return deps
+
+    def discover_callback_providers(self, callback: Callable
+                                    ) -> Set[PageObjectInputProvider]:
+        """Discover the providers that are required to fulfil the callback dependencies"""
+        plan = andi.plan(
+            callback,
+            is_injectable=is_injectable,
+            externally_provided=self.is_class_provided_by_any_provider,
+        )
+        result = set()
+        for cls, _ in plan:
```

Comment:
`provided_classes` can be a function, so it is not possible to just build a `dict` for speed up the check. This will be useful for implementing generic providers, for example, ones for `AutoExtract`.

```py
def provided_classes(cls):
  return issubclass(cls, AutoExtractPageInput)
```

I don't get your use of the word `priorities` here.

Reply:
As we have priorities at hand, we could create a `dict` with the highest priority provider, so there wouldn't be a need to iterate over things.

Reply:
Still, I don't get it. Could you please elaborate? The providers are already sorted by priority. But note that a provider could decide not to provide one of its classes listed, so we need the list of all of them, even if it has a lower priority. This is a rare feature, but it is as is implemented. Allows some providers to ignore providing some objects, for example, because of the URL. This allows creating a provider that overrides the default only for a domain for example, which can be useful in some situations.